frontend/appflowy_flutter/packages/appflowy_editor/lib/src/core/transform/transaction.dart

```diff
@@ -396,15 +404,23 @@ extension TextTransaction on Transaction {
               text,
             );
           } else {
-            var path = textNodes.first.path;
-            var j = i - textNodes.length + length - 1;
-            while (j > 0) {
-              path = path.next;
-              j--;
+            if (i == texts.length - 1) {
+              //ADD OFFSET CHARACTER TO END OF LAST TEXT-NODE TO AVOID DATA LOSS
+              document.insert(path, [
+                TextNode(
+                  delta: Delta()..insert("${text}${offSetChar}"),
+                ),
+              ]);
+            } else {
+              document.insert(path, [
+                TextNode(
+                  delta: Delta()..insert(text),
+                ),
+              ]);
             }
-            insertNode(path, TextNode(delta: Delta()..insert(text)));
```

Comment:
This was more complicated than I thought, existing insertNode function transformed paths, which in case of raise issue caused unexpected line skips and loss of data after selection in the same path (line), so I ended up using document.insert() which seemed to work fine.

Reply:
Actually, you should not use `document.insert` to update the document content. The only way to update it is through operations, like insertNode, deleteNode, and so on.

Reply:
Alright, I'll try to work it out.

Reply:
I got it to work thanks for review @LucasXu0